spring-web/src/main/java/org/springframework/http/client/reactive/ApacheClientHttpRequest.java

```diff
@@ -0,0 +1,142 @@
+import java.util.stream.Collectors;
+
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.message.BasicHttpRequest;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.core.io.buffer.DataBufferFactory;
+import org.springframework.http.HttpCookie;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+
+import static org.springframework.http.MediaType.ALL_VALUE;
+
+/**
+ * {@link ClientHttpRequest} implementation for the Apache HttpComponents HttpClient 5.x.
+ *
+ * @author Martin Tarjányi
+ * @since 5.3
+ * @see <a href="https://hc.apache.org/index.html">Apache HttpComponents</a>
+ */
+class ApacheClientHttpRequest extends AbstractClientHttpRequest {
+
+	private final HttpRequest httpRequest;
+
+	private final DataBufferFactory dataBufferFactory;
+
+	private Flux<ByteBuffer> byteBufferFlux;
+
+	private long contentLength = -1;
+
+
+	public ApacheClientHttpRequest(HttpMethod method, URI uri, DataBufferFactory dataBufferFactory) {
+		this.httpRequest = new BasicHttpRequest(method.name(), uri);
+		this.dataBufferFactory = dataBufferFactory;
+	}
+
+
+	@Override
+	public HttpMethod getMethod() {
+		return HttpMethod.resolve(this.httpRequest.getMethod());
+	}
+
+	@Override
+	public URI getURI() {
+		try {
+			return this.httpRequest.getUri();
+		}
+		catch (URISyntaxException ex) {
+			throw new IllegalArgumentException("Invalid URI syntax.", ex);
+		}
+	}
+
+	@Override
+	public DataBufferFactory bufferFactory() {
+		return this.dataBufferFactory;
+	}
+
+	@Override
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> body) {
+		return doCommit(() -> {
+			this.byteBufferFlux = Flux.from(body).map(DataBuffer::asByteBuffer);
+			return Mono.empty();
+		});
+	}
+
+	@Override
+	public Mono<Void> writeAndFlushWith(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+		return writeWith(Flux.from(body).flatMap(p -> p));
+	}
+
+	@Override
+	public Mono<Void> setComplete() {
+		return doCommit();
+	}
+
+	@Override
+	protected void applyHeaders() {
+		HttpHeaders headers = getHeaders();
+		this.contentLength = headers.getContentLength();
+
+		headers.entrySet()
+				.stream()
+				.filter(entry -> !HttpHeaders.CONTENT_LENGTH.equals(entry.getKey()))
+				.forEach(entry -> entry.getValue().forEach(v -> this.httpRequest.addHeader(entry.getKey(), v)));
+
+		if (!this.httpRequest.containsHeader(HttpHeaders.ACCEPT)) {
+			this.httpRequest.addHeader(HttpHeaders.ACCEPT, ALL_VALUE);
+		}
+	}
+
+	@Override
+	protected void applyCookies() {
+		if (getCookies().isEmpty()) {
+			return;
+		}
+
+		String cookiesString = getCookies().values()
+				.stream()
+				.flatMap(Collection::stream)
+				.map(HttpCookie::toString)
+				.collect(Collectors.joining("; "));
+
+		this.httpRequest.addHeader(HttpHeaders.COOKIE, cookiesString);
```

Comment:
I switched to using CookieStore. 

I have some concerns, though. CookieStore contains both request and response headers and it can even contain cookies from previous requests (depending on how the user configured the HttpClientContext). So when I [parse response cookies](https://github.com/spring-projects/spring-framework/blob/3b032e287c842ad3aba1d5580758e81354b09fd3/spring-web/src/main/java/org/springframework/http/client/reactive/HttpComponentsClientHttpResponse.java#L81) I have no way to tell which cookies came from the response and which ones are from a different source, so the current implementation returns all of them which is probably not the right behaviour(?).

Reply:
I understand your concern, but I think the potential for bugs in the former version was simply too high.

If the underlying implementation does not offer us a way to differentiate between various cookies, then so be it. If we document that behavior, then it's good enough for me.